src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs

```diff
@@ -132,9 +132,62 @@ public async Task<IServiceProvider> CreateContainerAsync(ShellSettings settings,
             // Rebuild the service provider from the updated collection.
             shellServiceProvider = tenantServiceCollection.BuildServiceProvider(true);
 
-            // Register all DIed types in ITypeFeatureProvider
             var typeFeatureProvider = shellServiceProvider.GetRequiredService<ITypeFeatureProvider>();
+            PopulateTypeFeatureProvider(typeFeatureProvider, featureAwareServiceCollection);
+
+            return shellServiceProvider;
+        }
+
+        private void EnsureApplicationFeature()
+        {
+            if (_applicationFeature is null)
+            {
+                lock (this)
```

Comment:
shouldn't the lock be for `_applicationFeature` instead of `this`? Then since `_applicationFeature` is locked, I don't think we need `??=` equals would be enough.